KAFKA-5854: Handle SASL authentication failures as non-retriable exceptions in clients

vahidhashemian opened the pull request:
This PR implements the client-side of KIP-152, by modifying `KafkaConsumer`, `KafkaProducer`, and `ConsumerGroupCommand` to throw a non-retriable exception when SASL authentication fails.

This PR is co-authored with @rajinisivaram.

tedyu commented on clients/src/main/java/org/apache/kafka/clients/KafkaClient.java:
```diff
@@ -68,6 +69,15 @@
+     * permanent and should not be resumed in the next {@link #ready(org.apache.kafka.common.Node, long)} } call.
+     *
+     * @param node the node to check
+     * @return true iff authentication has failed
```
Doesn't match return type of ApiException

vahidhashemian replied:
Thanks for catching this inconsistency. I'll fix it in the next patch.

vahidhashemian committed:
KAFKA-5854: Handle SASL authentication failures as non-retriable exceptions in clients

This PR implements the client-side of KIP-152, by modifying `KafkaConsumer`, `KafkaProducer`, and `ConsumerGroupCommand` to throw a non-retriable exception when user authentication fails.

This PR is co-authored with @rajinisivaram.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/ClusterConnectionStates.java:
```diff
@@ -136,6 +141,19 @@ public void ready(String id) {
     /**
+     * Enter the authentication failed state for the given node.
+     * @param id the connection identifier
+     * @param now the current time
```
missing param in javadoc `exception`

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/KafkaClient.java:
```diff
@@ -68,6 +69,15 @@
 
     /**
+     * Check if authentication to this node has failed, based on the connection state. Authentication failures are
+     * permanent and should not be resumed in the next {@link #ready(org.apache.kafka.common.Node, long)} } call.
```
That is not strictly true. e.g. producer.send() will throw an exception on an authentication failure, but the produce is still usable. You can invoke `producer.send()` again to retry if you wish to. So it may be better to say that authentication failures are propagated without any retries, without saying `permanent`.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -312,6 +313,18 @@ public boolean connectionFailed(Node node) {
 
     /**
+     * Check if authentication to this node has failed, based on the connection state. Authentication failures are
+     * permanent and should not be resumed in the next {@link #ready(org.apache.kafka.common.Node, long)} } call.
```
Same as previous comment. Failures are not `permanent`, but no retries are performed.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -588,6 +601,7 @@ private void processDisconnection(List<ClientResponse> responses, String nodeId,
         nodesNeedingApiVersionsFetch.remove(nodeId);
         switch (disconnectState.state()) {
             case AUTHENTICATION_FAILED:
+                connectionStates.authenticationFailed(nodeId, now, Errors.AUTHENTICATION_FAILED.exception());
```
This should be `connectionStates.authenticationFailed(nodeId, now, disconnectState.exception())`. That contains the exception message provided by the broker and we want to propagate that to the user.

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -596,6 +610,7 @@ private void processDisconnection(List<ClientResponse> responses, String nodeId,
                         "that authentication failed due to invalid credentials.", nodeId);
                 break;
             case NOT_CONNECTED:
+                connectionStates.disconnected(nodeId, now);
```
This is already done at the start of the method, so not required here?

vahidhashemian replied:
That's right. Thanks!

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:
```diff
@@ -463,6 +468,26 @@ public boolean connectionFailed(Node node) {
+     * Find whether a previous connection has succeeded.
+     * @param node Node to connect to if possible
+     */
+    public boolean isReady(Node node) {
```
Not sure if this method is required at all. Ready is a transient state, so it will be difficult to use in a meaningful way.

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslPlainPlaintextClientsWithInvalidCredentialsTest.java:
```diff
@@ -0,0 +1,133 @@
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class SaslPlainPlaintextClientsWithInvalidCredentialsTest {
```
Perhaps this could just be named `ClientAuthenticationFailureTest`? We could add other protocols in future to the same test. Also it should perhaps be in the `clients` package rather than the common security package since it is testing the client level behavior.

vahidhashemian replied:
I tried moving it to the `clients` package but due to the following imports that seem necessary I am getting `checkstyle` errors that I could not find a way to resolve. For this reason, I kept it in the same place.
```
import org.apache.kafka.clients.consumer.ConsumerConfig;
import org.apache.kafka.clients.consumer.KafkaConsumer;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerConfig;
import org.apache.kafka.clients.producer.ProducerRecord;
```

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslPlainPlaintextClientsWithInvalidCredentialsTest.java:
```diff
@@ -0,0 +1,133 @@
+        try (KafkaConsumer<String, String> consumer = new KafkaConsumer<>(saslClientConfigs)) {
+            consumer.subscribe(Arrays.asList(topic));
+            consumer.poll(100);
+            fail("Expeceted an authentication error!");
```
typo in `Expected`

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslPlainPlaintextClientsWithInvalidCredentialsTest.java:
```diff
@@ -0,0 +1,133 @@
+        } catch (AuthenticationFailedException e) {
+            // OK
+        } catch (Exception e) {
+            fail("Expecetd only an authentication error, but another error occurred: " + e.getMessage());
```
typo in `Expected`

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/security/authenticator/SaslPlainPlaintextClientsWithInvalidCredentialsTest.java:
```diff
@@ -0,0 +1,133 @@
+        propsStream.write("  password=\"anotherpassword\";\n".getBytes());
+        propsStream.write("};".getBytes());
+        propsStream.close();
+        System.setProperty("java.security.auth.login.config", jaasConfigFile.getAbsolutePath());
```
You can replace this block of code creating JaasConfig with
```
 testJaasConfig = TestJaasConfig.createConfiguration("PLAIN", Arrays.asList("PLAIN"));
 testJaasConfig.setClientOptions("PLAIN", TestJaasConfig.USERNAME, "anotherpassword");
```

rajinisivaram commented on core/src/main/scala/kafka/admin/AdminClient.scala:
```diff
@@ -52,15 +52,28 @@ class AdminClient(val time: Time,
+              authenticationPerformed = false
+          }
+          if (authenticationPerformed)
+            nonRetriableException = None
```
I dont think we need to check or remember if authentication failed. Instead we should propagate AuthenticationExceptions from `AdminClient.sendAnyNode` to avoid retrying. I think this block of change can be removed.

vahidhashemian replied:
Thanks. I'll use your suggested approach.

rajinisivaram commented on core/src/main/scala/kafka/admin/AdminClient.scala:
```diff
@@ -52,15 +52,28 @@ class AdminClient(val time: Time,
                   val bootstrapBrokers: List[Node]) extends Logging {
 
   @volatile var running: Boolean = true
+  @volatile var nonRetriableException: Option[ApiException] = null
```
Don't think this is required.

rajinisivaram commented on core/src/main/scala/kafka/admin/AdminClient.scala:
```diff
@@ -78,6 +91,8 @@ class AdminClient(val time: Time,
   networkThread.start()
 
+  def getNonRetriableException = this.nonRetriableException
+
```
Don't think this is required.

rajinisivaram commented on core/src/main/scala/kafka/admin/AdminClient.scala:
```diff
@@ -478,11 +493,20 @@ object AdminClient {
+
+    if (adminClient.getNonRetriableException.isDefined)
+      throw adminClient.getNonRetriableException.get
+
```
This is the reason for the transient timeout test failure. This waits for a successful connection, which causes the first request to succeed in the test. This block of code is not required. We should simply propagate authentication exceptions from request failures.

rajinisivaram commented on core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala:
```diff
@@ -419,6 +419,8 @@ object ConsumerGroupCommand extends Logging {
     private var consumer: KafkaConsumer[String, String] = null
 
+    def getNonRetriableException = adminClient.getNonRetriableException
+
```
Not required.

rajinisivaram commented on core/src/test/scala/integration/kafka/api/SaslPlainPlaintextClientsWithInvalidCredentialsTest.scala:
```diff
@@ -0,0 +1,122 @@
+import kafka.server.KafkaConfig
+import kafka.utils.{JaasTestUtils, TestUtils}
+
+class SaslPlainPlaintextClientsWithInvalidCredentialsTest extends BaseConsumerTest with SaslSetup {
```
Would it be possible to avoid extending `BaseConsumerTest`? This would run all tests from `BaseConsumerTest`, which we probably don't want to do? If it is too big a change, then the current overriding of methods may be ok.

rajinisivaram commented on core/src/test/scala/integration/kafka/api/SaslPlainPlaintextClientsWithInvalidCredentialsTest.scala:
```diff
@@ -0,0 +1,122 @@
+   * when zookeeper.set.acl=false, even if Zookeeper is SASL-enabled.
+   */
+  @Test
+  def testZkAclsDisabled() {
```
Is this test required here? It is bringing up a cluster and tearing it down, which is not required for this test class?

vahidhashemian replied:
Correct, it's not needed here. I'll remove it.

rajinisivaram commented on core/src/test/scala/integration/kafka/api/SaslSetup.scala:
```diff
@@ -100,6 +100,24 @@ trait SaslSetup {
+      case Both => Seq(JaasTestUtils.kafkaServerSection(kafkaServerEntryName, kafkaServerSaslMechanisms, serverKeytabFile),
+        JaasTestUtils.kafkaClientSectionInvalidCredentials(kafkaClientSaslMechanism, clientKeytabFile)) ++ JaasTestUtils.zkSections
+    }
+  }
```
This method is not required. For invalid client credentials, you can simply override `sasl.jaas.config` property

rajinisivaram reviewed (commented):
@vahidhashemian Thank you for the PR. I have left some suggestions. I think the tests you have added cover error propagation with all the clients, which is good. It will also be useful to have tests which check that if the configuration becomes valid after a failure, the clients can reconnect successfully. I could create another JIRA and help with that to try and get this one merged soon since we are so close to code freeze. Let me know what you think.

rajinisivaram commented:
@vahidhashemian I wrote a different test using SCRAM rather than PLAIN since SCRAM credentials can be dynamically updated and also made a few fixes to the code. This covers the additional test after credential update I had mentioned earlier. My changes are in https://github.com/rajinisivaram/kafka/tree/KAFKA-5854-KIP-152-client. Can you take a look and merge any changes you find useful into your branch. And once you have addressed the review comments so far, it will be good to get another opinion on the general approach. Can you ping @hachikuji  for a review when it is ready? Thank you!

vahidhashemian commented:
@rajinisivaram Thanks for the earlier review and the additional tests. I'll try to address the comments and merge your tests into this PR today.

vahidhashemian committed:
Addressed feedback on the initial commit.

Thanks to @rajinisivaram for the review and sharing some of the code to speed things up.

vahidhashemian reviewed (commented):
Thanks @rajinisivaram for the feedback and the suggested fixes. I submitted an update to address the feedback.

cc @hachikuji

rajinisivaram commented:
retest this please

rajinisivaram commented:
@vahidhashemian Thank you for the updates. I have started a system test run on your branch to ensure the tests pass.
@hachikuji Can you also do a high level review of the implementation changes to make sure I haven't missed anything? Thank you!

rajinisivaram commented:
@vahidhashemian we need to update `KafkaAdminClient` as well. I could do that in a separate PR if you prefer, depending on how busy you are. Let me know what you think. Thanks.

rajinisivaram commented:
@vahidhashemian Can you update javadoc for the public APIs (e.g. `KafkaConsumer.poll()`, `KafkaProducer.send()` etc.) to include `AuthenticationException` in the `throws` section? Thank you!

vahidhashemian commented:
@rajinisivaram I'll work on updating `KafkaAdminClient` and also updating the javadoc today, and let you know how it goes. Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/ConnectionState.java:
```diff
@@ -16,14 +16,21 @@
-    DISCONNECTED, CONNECTING, CHECKING_API_VERSIONS, READY
+    DISCONNECTED, CONNECTING, CHECKING_API_VERSIONS, READY, AUTHENTICATION_FAILED;
+
+    public static EnumSet<ConnectionState> disconnectedStates() {
```
Another option is to add an `isDisconnected` field. Then the check is just `state.isDisconnected` instead of `disconnectedStates().contains(state)`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/NetworkClient.java:
```diff
@@ -847,6 +864,13 @@ public void handleDisconnection(String destination) {
         }
 
         @Override
+        public void handleAuthenticationFailure(ApiException exception) {
```
Hmm.. I wonder if we may as well use the more specific `AuthenticationFailedException` rather than `ApiException` everywhere.

vahidhashemian replied:
To remain consistent with the exception type used in `KafkaChannel.java` I replaced `ApiException` with `AuthenticationException` (which is a super-type of `AuthenticationFailedException`).

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -59,6 +60,7 @@
     private int version;
     private long lastRefreshMs;
     private long lastSuccessfulRefreshMs;
+    private ApiException nonRetriableException;
```
From a quick glance at the code, the only case where this is populated is an authentication failure. If so, maybe we can give it a more specific name?

vahidhashemian replied:
I changed it to `nonRetriableAuthenticationException`. I hope that works better.

hachikuji reviewed (commented):
Starting to look this over. Left a couple quick comments.

rajinisivaram commented:
@vahidhashemian Can you do the javadoc and the updates suggested by @hachikuji first so that I can try and merge this PR tomorrow? If you have time remaining, you could do the `KafkaAdminClient` changes as a separate PR today and I could review it tomorrow. Otherwise, I could do that one tomorrow morning and you could help review it when you get in. Thanks.

vahidhashemian commented:
@rajinisivaram Sure, sounds good. I'll work on those feedback and submit a PR shortly.

vahidhashemian committed:
Addressed additional feedback received

vahidhashemian reviewed (commented):
@hachikuji Thanks for the feedback. I responded to some inline, and updated the code accordingly in the new commit.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -59,6 +60,7 @@
     private int version;
     private long lastRefreshMs;
     private long lastSuccessfulRefreshMs;
+    private AuthenticationException nonRetriableAuthenticationException;
```
Are there retriable authentication errors?

vahidhashemian replied:
Not at this point. I'll make the name more concise for now: `authenticationException`.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/ConnectionState.java:
```diff
@@ -23,7 +23,12 @@
+    DISCONNECTED, CONNECTING, CHECKING_API_VERSIONS, READY, AUTHENTICATION_FAILED;
+
+    public boolean isDisconnected() {
+        return this.equals(AUTHENTICATION_FAILED) || this.equals(DISCONNECTED);
```
nit: you could use `==`

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/ManualMetadataUpdater.java:
```diff
@@ -69,6 +70,11 @@ public void handleDisconnection(String destination) {
 
     @Override
+    public void handleAuthenticationFailure(AuthenticationException exception) {
+        // Do nothing
```
Not sure if it is in scope, but are we not handling authentication failures between brokers? Maybe we should at least log a warning or something?

vahidhashemian replied:
I'll add a warning here.

rajinisivaram replied:
We don't fail the broker on authentication failures, but there is sufficient information in the broker logs to identify the failure. I think this log entry should be a `debug` rather than a `warn` since we already have an error log entry, so the warning doesn't add much value. With this change the log looks like:
```
ERROR [Controller id=0, targetBrokerId=0] Connection to node 0 failed authentication due to: Authentication failed due to invalid credentials with SASL mechanism PLAIN (org.apache.kafka.clients.NetworkClient)
WARN An authentication error occurred in broker-to-broker communication! (org.apache.kafka.clients.ManualMetadataUpdater)
```

ijuma replied:
Agreed. And we should add a comment explaining that.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -146,17 +148,32 @@ public synchronized boolean updateRequested() {
+    public synchronized void maybeHandleNonRetriableAuthenticationExceptions() {
+        if (nonRetriableAuthenticationException != null) {
+            AuthenticationException exception = nonRetriableAuthenticationException;
+            this.needUpdate = false;
```
Curious what the reason for this is.

vahidhashemian replied:
cc @rajinisivaram

rajinisivaram replied:
It is to avoid unnecessary connections being made to update metadata e.g. from `Sender` thread which will continue to retry if `needUpdate` is set. The flag will be set again if the application performs another operation (e.g. another send), but we are avoiding automatic retry.

hachikuji replied:
I see. My concern is that we might lose track of the fact that we still need the metadata update. For example, we request a metadata when the user calls `consumer.subscribe(Pattern)`. If we then raise an `AuthenticationException` from `poll()` and the user ignores it, then we might not bother attempting another metadata update until the next refresh time. Maybe that is an acceptable tradeoff to avoid having the background threads retrying needlessly until shutdown? At a minimum, it would be useful to have a comment here explaining the purpose.

rajinisivaram replied:
@hachikuji Both producers and consumers request metadata update again when waiting for metadata or polling in the consumer. The integration tests do check that the operation succeeds after credentials are created. Of course, a comment is useful, @vahidhashemian can you add one? More tests using `MockClient` would also be be useful, but we could do that in another PR.

hachikuji replied:
Hmm.. In the consumer, we only fetch metadata if we know we need to, and that is sometimes conveyed only through the `needUpdate` flag (as in the case I mentioned).

One additional note: this flag alone is not sufficient to prevent metadata retries if the metadata max age has expired (which will be the case on initialization).

rajinisivaram replied:
@hachikuji  I may have missed something, but in the consumer case, we have no coordinator or leader information since we could not fetch metadata. So we will have to refresh on the next `poll()`. Both `ensureCoordinatorReady` and the `Fetcher` request updates.

On the second note: I was thinking that we would clear the topics on authentication exception in producer to avoid automatic refresh on expiry, but I had forgotten about that. Perhaps we should? On the consumer, I was thinking this is not an issue since you have to explicitly `poll()` to trigger an update - have I missed something?

hachikuji replied:
@rajinisivaram Yeah, I think we are handling the common (initialization) cases in this patch, just a bit discomforting to leave some of the other cases a little nebulously handled. 

In this particular case, it doesn't really seem worth trying to prevent additional metadata retries since the NetworkClient will still enforce a backoff and we expect the user to just close the client anyway. I would probably leave `needUpdate` untouched as we do currently on failure.

rajinisivaram replied:
@hachikuji OK, I am fine with that. @vahidhashemian Can you remove the `needUpdate` change here? Thanks.

vahidhashemian replied:
Sure. Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/network/ChannelState.java:
```diff
@@ -86,7 +86,7 @@ public State state() {
+    public AuthenticationException exception() {
         return exception;
     }
 }
```
nit: missing newline

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:
```diff
@@ -134,6 +135,7 @@ public boolean awaitMetadataUpdate(long timeout) {
         int version = this.metadata.requestUpdate();
         do {
             poll(timeout);
+            this.metadata.maybeHandleNonRetriableAuthenticationExceptions();
```
Hmm.. What happens for metadata updates that occur without an explicit call to `awaitMetadataUpdate`? It could be triggered by the `NetworkClient` itself after all. Maybe we need to add something to the consumer's implementation of `Metadata.Listener`?

vahidhashemian replied:
This piece covers authentication failure handling for when partitions are manually assigned.
The case where partitions are auto-assigned is handled in the `ensureCoordinatorReady` method (via `throw future.exception()` call). Could there be other authentication failure scenarios for consumers we are not handling?
cc @rajinisivaram

rajinisivaram replied:
@hachikuji The scope of the JIRA and this PR are limited to a specific scenario - provide better feedback to applications when security is misconfigured. We currently retry forever, instead we want to throw an `AuthenticationException`. To do this, the first request initiated by any public API needs to process authentication exceptions. And I think that is covered by this PR for producers and consumers.
There is a second scenario, which we do want to handle eventually, but is not addressed by this PR. And that is addressing proper handling of credential updates. If a credential is removed, then we could see authentication exception from any request and we want to propagate this properly. This needs quite extensive testing and is less likely to be hit by users so it is not being addressed now.

ijuma replied:
Btw, we can perhaps do that in a separate PR, but the AdminClient also needs to handle this correctly (not just Producer and Consumer). Also, we should file a JIRA for Streams and Connect to verify that they behave correctly too.

rajinisivaram replied:
@ijuma Yes, I had asked Vahid to do `AdminClient` in a separate PR so that we can get this one committed. He started on it yesterday and has some initial code for review. For Connect and Streams, I will file a separate JIRA. There is another missing piece which is the transactional API with producers, I will do that in another PR now.

hachikuji replied:
@vahidhashemian Hmm.. For manual partition assignment, there is no guarantee that the `ensureCoordinatorReady` path will even be executed. We intentionally avoid looking up the coordinator if the user manually sets their own offsets and does not use the commit APIs. Maybe we can add an integration test for this path to ensure that the `AuthenticationException` is somehow getting propagated. Basically it should turn off auto-commit, assign a partition, seek to offset 0 (say), and then call `poll()`.

rajinisivaram replied:
@vahidhashemian It should be easy to add the scenario @hachikuji has suggested above by adding another test similar to `SaslClientsWithInvalidCredentialsTest.testManualAssignmentConsumerWithAuthenticationFailure`.

vahidhashemian replied:
Sure, I'll try to add that scenario. Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -146,17 +148,32 @@ public synchronized boolean updateRequested() {
+     * If any non-retriable authentication exceptions were encountered during
+     * metadata update, clear and throw the exception.
+     */
+    public synchronized void maybeHandleNonRetriableAuthenticationExceptions() {
```
How about `maybeThrowAuthenticationException`?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1053,6 +1053,7 @@ public void assign(Collection<TopicPartition> partitions) {
      *             function is called
      * @throws org.apache.kafka.common.errors.InterruptException if the calling thread is interrupted before or while
      *             this function is called
+     * @throws org.apache.kafka.common.errors.AuthenticationFailedException if caller's authentication fails
```
Is this type correct? It seems like we're only sure that it is an instance of `AuthenticationException`. Also, is `poll()` the only API that can receive this? It seems like it should be possible from any of the blocking APIs.

vahidhashemian replied:
Thanks for catching this. I missed updating this when I changed the types in the last commit. Will also update javadocs of other affected methods.

rajinisivaram replied:
A few methods are still missing `AuthenticationException` (e.g. `partitionsFor` in producer and consumer)

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java:
```diff
@@ -367,9 +369,13 @@ private void checkDisconnects(long now) {
-                        null, null));
+                    AuthenticationException authenticationException = client.authenticationException(node);
+                    if (authenticationException !=  null)
+                        handler.onFailure(authenticationException);
```
I guess it's possible that we could get this error from the consumer's heartbeat thread. If so, we will swallow it. Same thing if we get a `GroupAuthorizationException` I guess. We actually do have a way to propagate exceptions from the heartbeat thread to the `poll()`. Maybe we should try to handle these two cases?

vahidhashemian replied:
I added handlers for these two error cases.

hachikuji reviewed (commented):
Thanks for the patch. Left a few more comments.

vahidhashemian committed:
Minor changes based on the feedback received

vahidhashemian reviewed (commented):
@hachikuji Thanks again for the feedback. I tried to address most of them in the new commit. I cc'ed @rajinisivaram on a couple of your comments as I feel she can better respond to those.

Please let me know if I missed anything. Thanks.

vahidhashemian committed:
Merge branch 'trunk' into KAFKA-5854

ijuma commented on clients/src/main/java/org/apache/kafka/clients/ManualMetadataUpdater.java:
```diff
@@ -71,7 +75,7 @@ public void handleDisconnection(String destination) {
     @Override
     public void handleAuthenticationFailure(AuthenticationException exception) {
-        // Do nothing
+        log.warn("An authentication error occurred in broker-to-broker communication!", exception);
```
This should probably be an `error`.

ijuma replied:
I retract this based on Rajini's comment that we log the error elsewhere.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1053,7 +1053,7 @@ public void assign(Collection<TopicPartition> partitions) {
      * @throws org.apache.kafka.common.errors.InterruptException if the calling thread is interrupted before or while
      *             this function is called
-     * @throws org.apache.kafka.common.errors.AuthenticationFailedException if caller's authentication fails
+     * @throws org.apache.kafka.common.errors.AuthenticationException if caller's authentication fails
```
I noticed that there is no Javadoc for AuthenticationException and subclasses. It would be good to add some and to explain what is the difference between them. It could be done in a separate PR. cc @rajinisivaram

ijuma replied:
Same for AuthorizationException.

rajinisivaram replied:
@ijuma Thank you for the review. @vahidhashemian Can you add `@see AuthenticationException/AuthorizationException` to all the javadoc entries and I will submit another PR with more docs for the exceptions. Thanks.

ijuma commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1166,6 +1166,7 @@ public boolean shouldBlock() {
      *             function is called
      * @throws org.apache.kafka.common.errors.InterruptException if the calling thread is interrupted before or while
      *             this function is called
+     * @throws org.apache.kafka.common.errors.AuthenticationException if caller's authentication fails
```
Nit: maybe "if authentication failed, see AuthenticationException for more details". And the javadoc for AuthenticationException can include a bit more information of the reasons, etc.

ijuma reviewed (commented):
I had a very quick look and left some minor comments.

vahidhashemian committed:
Javadoc updates

vahidhashemian commented:
@ijuma @rajinisivaram Thanks for the feedback. Just submitted another update.

vahidhashemian committed:
Changed @see javadoc tag to @link

rajinisivaram commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -146,17 +148,32 @@ public synchronized boolean updateRequested() {
+     * If any non-retriable authentication exceptions were encountered during
+     * metadata update, clear and throw the exception.
+     */
+    public synchronized void maybeThrowAuthenticationExceptions() {
```
Plural not required here: `maybeThrowAuthenticationException`?

rajinisivaram commented on clients/src/test/java/org/apache/kafka/clients/MockClient.java:
```diff
@@ -135,6 +136,12 @@ public boolean connectionFailed(Node node) {
 
     @Override
+    public AuthenticationException authenticationException(Node node) {
+        //TODO
```
TODO can be removed since we are not using this in tests?

hachikuji replied:
Maybe we should add a JIRA to implement this? Seems like a testing gap if we are not needing it since most of the consumer unit tests depend on `MockClient`. We should have test cases to ensure the behavior if there is an authentication failure in any of the various APIs.

rajinisivaram replied:
Yes, we should have a JIRA to implement unit tests using the `MockClient`. The tests in this PR do actual authentication. @vahidhashemian Would you like to work on that in a separate PR?

rajinisivaram replied:
Opened another JIRA for mock tests: https://issues.apache.org/jira/browse/KAFKA-5944

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/security/authenticator/ClientAuthenticationFailureTest.java:
```diff
@@ -0,0 +1,114 @@
+        } catch (AuthenticationFailedException e) {
+            // OK
+        } catch (Exception e) {
+            fail("Expecetd only an authentication error, but another error occurred: " + e.getMessage());
```
typo in Expected

rajinisivaram commented on clients/src/test/java/org/apache/kafka/common/security/authenticator/ClientAuthenticationFailureTest.java:
```diff
@@ -0,0 +1,114 @@
+        try (KafkaConsumer<String, String> consumer = new KafkaConsumer<>(saslClientConfigs)) {
+            consumer.subscribe(Arrays.asList(topic));
+            consumer.poll(100);
+            fail("Expeceted an authentication error!");
```
typo in `Expeceted`

rajinisivaram reviewed (commented):
@vahidhashemian Thanks for the updates. Left a few more very minor comments.
@hachikuji Can you take a quick look to see if your concerns have been addressed? I will merge once all your feedback is addressed. Thank you!

vahidhashemian committed:
Fixed typos

vahidhashemian commented:
@rajinisivaram Thanks for catching those typos. I fixed them in the new commit.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:
```diff
@@ -1379,8 +1382,9 @@ public void seekToEnd(Collection<TopicPartition> partitions) {
+     * @throws org.apache.kafka.common.errors.AuthenticationException if authentication fails. {@link AuthenticationException}
      * @throws org.apache.kafka.common.errors.AuthorizationException if not authorized to the topic or to the
-     *             configured groupId
+     *             configured groupId. {@link AuthorizationException}
```
Are all of these links actually necessary? From what I can tell, the link is provided automatically by the `@throws`. See here: https://kafka.apache.org/0110/javadoc/index.html?org/apache/kafka/clients/consumer/KafkaConsumer.html.

vahidhashemian replied:
Hmm, I see similar `@link` tags in the javadoc. For example, `Deserializer` and `ConsumerConfig` are hyperlinked for the constructor method [here](https://kafka.apache.org/0110/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#KafkaConsumer(java.util.Map,%20org.apache.kafka.common.serialization.Deserializer,%20org.apache.kafka.common.serialization.Deserializer)), and there are corresponding `@link` tags in the source code [here](https://github.com/apache/kafka/blob/bd0146d984dd5df82fb19aa936e8f4ff9ca40030/clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java#L590).

vahidhashemian replied:
Sorry, you're right. It seems the `@throws` tag automatically creates a link. I'll remove them. @ijuma @rajinisivaram please advise if you see an issue with removing them.

ijuma replied:
Sorry, I didn't mean it like this. I simply meant that we should emphasize that details of when AuthenticationException is thrown are described there. Something as simple as "See the exception for more details" is enough. This should be implicit, but lack of docs is so common that people often don't think to look.

vahidhashemian replied:
Thinking about this a bit more, wouldn't we need to make a similar statement for (almost) all `@throws` tags? I'm fine taking either approach, but I think we should be consistent (as much as possible).

ijuma replied:
The reason why it's worth it in this case is that this is a complicated area where we get a lot of questions. The consumer documents in detail many of the potential exceptions that are specific to the consumer, but AuthenticationException applies equally to all clients so it makes sense to centralise the docs in the exception and nudge users in that direction.

vahidhashemian replied:
OK, that's fair. I'll add it for `AuthenticationException` and `AuthorizationException`. Thanks.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java:
```diff
@@ -982,6 +983,12 @@ public void onFailure(RuntimeException e) {
                         }
                     }
                 }
+            } catch (AuthenticationException e) {
```
I'm not sure this is the right place for this check. I think it would have to be in the `onFailure` hook above. A test case might clarify this.

rajinisivaram replied:
@hachikuji Since heartbeat thread is started only after a join request succeeds, this is perhaps not needed for this PR?

hachikuji replied:
Yes, if we are only concerned with authentication errors during initialization, then we can leave it for later. We should make sure to capture the gaps in another JIRA though.

rajinisivaram replied:
Created https://issues.apache.org/jira/browse/KAFKA-5945

vahidhashemian committed:
Added an additional integration test and updated javadocs

vahidhashemian commented:
I think the most recent commit addresses the outstanding comments received (to be done as part of this PR). Please advise if something is amiss.

vahidhashemian committed:
Minor update to address the feedback

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/Metadata.java:
```diff
@@ -146,17 +148,31 @@ public synchronized boolean updateRequested() {
-            if (remainingWaitMs != 0)
+            if (remainingWaitMs != 0) {
                 wait(remainingWaitMs);
+                maybeThrowAuthenticationException();
```
Might not a big problem, but I wonder if we should check for the authentication exception before the `wait` as well? It is possible that `awaitUpdate` returns before the authentication failure happens. A subsequent call may then begin with the `authenticationException` not null which would cause a needless `wait`. I think an easy solution is to move this line up to the beginning of the `while` block.

vahidhashemian replied:
Thanks for the suggestion. I moved it to the top of the `while` block, outside the `if` block to give preference to authentication exception over timeout exception in the edge case.

hachikuji reviewed (approved):
One more minor comment, but the rest LGTM.

vahidhashemian committed:
Avoid a potential metadata wait by checking for an authentication error first

rajinisivaram reviewed (approved):
@vahidhashemian Thank you very much for your patience with this PR. LGTM.

rajinisivaram commented:
Merging to trunk.
